tmpl/wiki/mirage2-to-mirage3.md

```diff
@@ -0,0 +1,1416 @@
+val default_posix_clock: pclock impl
+val default_monotonic_clock: mclock impl
+val default_reporter:
+val default_random: random impl
+val default_console: console impl
+val default_io_page: io_page impl
+val generic_kv_ro:
+val default_network: network impl
+val generic_stackv4:
+val default_argv: Functoria_app.argv impl
+```
+
+Many existing unikernels will find it useful to replace their existing network stack detecting code with `generic_stackv4` in particular -- let's see an example.
+
+## porting the stackv4 example to Mirage 3
+
+A good example of replacing old configurator code can be seen between [the pre-functoria `network` example](https://github.com/mirage/mirage-skeleton/tree/21adfc85b124e886d871079f28bd0a868ba3c5fb/stackv4), and [the `device-usage/network` example from the Mirage-3.0.0-compatible branch of `mirage-skeleton`](https://github.com/mirage/mirage-skeleton/tree/e9360fa1ce02d26a8931238a16000fa12df40ebf/device-usage/network).  We can replace this old code from `config.ml`:
+
+```
+let net =
+  try match Sys.getenv "NET" with
+    | "direct" -> `Direct
+    | "socket" -> `Socket
+    | _        -> `Direct
+  with Not_found -> `Direct
+
+let dhcp =
+  try match Sys.getenv "ADDR" with
+    | "dhcp"   -> `Dhcp
+    | "static" -> `Static
+  with Not_found -> `Dhcp
+
+let stack console =
+  match net, dhcp with
+  | `Direct, `Dhcp   -> direct_stackv4_with_dhcp console tap0
+  | `Direct, `Static -> direct_stackv4_with_default_ipv4 console tap0
+  | `Socket, _       -> socket_stackv4 console [Ipaddr.V4.any]
+
+```
+
+with this:
+
+```
+let stack = generic_stackv4 default_network
+```
+
+You may notice that `generic_stackv4` doesn't take a `console` argument, where the previous `direct_stackv4*` and `socket_stackv4` functions did.  This leads us to another big change that predates MirageOS 3, but is more widely used and better supported in this release: logging.
+
+The `mirage_logs` library and its incorporation into the configuration language dates back to version 2.9.0, but a number of libraries still required a `console impl` argument and expected to log to the console.  In Mirage 3.0.0, we've tried to replace the calls to `Console.log` with calls to `Logs.debug`, `Logs.info`, or `Logs.warn` as appropriate.  Consequently, many of the functions which previously needed to be passed a console no longer require it.  For example, we can take [the `stackv4` example from `mirage-skeleton` commit f36d2958f616fb882df37f08d3440797471ca0cc](https://github.com/mirage/mirage-skeleton/tree/f36d2958f616fb882df37f08d3440797471ca0cc/stackv4), which fails with Mirage 3:
+
+```
+mirage-skeleton/stackv4$ cat config.ml
+open Mirage
+
+let handler = foreign "Unikernel.Main" (console @-> stackv4 @-> job)
+
+let stack = generic_stackv4 default_console tap0
+
+let () =
+  register "stackv4" [handler $ default_console $ stack]
+mirage-skeleton/stackv4$ 
+mirage-skeleton/stackv4$ mirage configure -t unix
+mirage: unknown option `-t'.
```

Comment:
This is a bit confusing. Wouldn't mirage stop after getting an unknown option? Cut-and-paste error?